crates/fluvio-protocol-derive/src/util.rs

```diff
@@ -1,75 +1,273 @@
+        )),
+    }?;
+
+    // to_string so we can return the name of the ident
+    // helps to reduce the repetitiveness of attribute names
+    let attr_name = ident.to_string();
+    let attr_span = ident.span();
+
+    if let Ok(value) = meta.value() {
+        if let Ok(expr) = value.parse() {
+            return Ok((Some(expr), attr_span, attr_name));
+        }
+    }
+
+    Ok((None, attr_span, attr_name))
 }
 
-pub(crate) fn find_meta<'a>(meta: &'a Meta, name: &str) -> Option<&'a Meta> {
-    if let Meta::List(list) = meta {
-        for attr in list.nested.iter() {
-            if let NestedMeta::Meta(named_meta) = attr {
-                match named_meta {
-                    Meta::NameValue(meta_name_value) => {
-                        if meta_name_value.path.is_ident(name) {
-                            return Some(named_meta);
-                        }
-                    }
-                    Meta::Path(path) => {
-                        if path.is_ident(name) {
-                            return Some(named_meta);
-                        }
+pub(crate) use parse_attributes;
+
+pub fn get_attr_type_from_meta(meta: &ParseNestedMeta) -> syn::Result<PropAttrsType> {
+    let (expr, attr_span, attr_name) = parse_attributes_data(meta)?;
+    get_expr_value(&attr_name, expr.as_ref(), attr_span)
+}
+pub fn get_expr_value<'a>(
+    attr_name: &'a str,
+    field: Option<&'a Expr>,
+    span: Span,
+) -> syn::Result<PropAttrsType> {
+    match &field {
+        Some(Expr::Lit(lit_expr)) => {
+            if let Lit::Int(lit) = &lit_expr.lit {
+                Ok(PropAttrsType::Int(lit.base10_parse::<i16>()?))
+            } else if let Lit::Str(lit) = &lit_expr.lit {
+                let value = &lit.value();
+
+                // To handle functions passed as lit we check for ()
+                // And strip it when creating the new Ident so we can use it later on
+                if value.contains('(') && value.contains(')') {
+                    if let Some(value) = &lit.value().strip_suffix("()") {
+                        return Ok(PropAttrsType::Fn(syn::Ident::new(value.trim(), lit.span())));
                     }
-                    Meta::List(_) => {}
                 }
+
+                Ok(PropAttrsType::Lit(syn::Ident::new(value, lit.span())))
+            } else {
+                Err(syn::Error::new(
+                    span,
+                    format!("Expected {attr_name} attribute to be a LitStr or LitInt: `{attr_name} = \"...\"`"),
+                ))
             }
         }
+        Some(Expr::Unary(ExprUnary { expr, .. })) => {
+            // When passing -1 as a value it is returned as type Unary
+            // So to handle that we are checking if it's Unary Lit and continue as usual
+            // If needed this can be extended to handle the Unary operators
+            // But it doesn't seem that is necessary currently
+            Ok(PropAttrsType::Int(-parse_int_expr(attr_name, expr, span)?))
```

Comment:
`parse_int_expr` is also handling `unary` expr inside. We should decide on which level we want to deal with it and do it only once.

Reply:
They do both handle Unary, but in this because we are in the Unary case the type of `expr` is `Expr::Lit` so just using the `parse_int_expr` to get the Literal from there and convert it so we save some lines of code and set it as a negative because we know this is Unary. Of course in the future if needed we can use the other operators as well and the logic might change but i don't see a reason to do it for now as the only place Unary is used if -1 is being passed.